packages/react-components/react-virtualizer/src/useVirtualizer.ts

```diff
@@ -0,0 +1,424 @@
+    const iBefore = Math.max(midpoint - 1, 0);
+    const iAfter = Math.min(midpoint + 1, childProgressiveSizes.current.length - 1);
+    const indexValue = childProgressiveSizes.current[midpoint];
+    const afterIndexValue = childProgressiveSizes.current[iAfter];
+    const beforeIndexValue = childProgressiveSizes.current[iBefore];
+    if (scrollPos <= afterIndexValue && scrollPos >= beforeIndexValue) {
+      /* We've found our index - if we are exactly matching before/after index that's ok,
+      better to reduce checks if it's right on the boundary. */
+      return midpoint;
+    }
+
+    if (indexValue > scrollPos) {
+      return findIndexRecursive(scrollPos, lowIndex, midpoint - 1);
+    } else {
+      return findIndexRecursive(scrollPos, midpoint + 1, highIndex);
+    }
+  };
+
+  const getIndexFromSizeArray = (scrollPos: number): number => {
+    /* Quick searches our progressive height array */
+    if (
+      scrollPos === 0 ||
+      childProgressiveSizes.current.length === 0 ||
+      scrollPos <= childProgressiveSizes.current[0]
+    ) {
+      // Check start
+      return 0;
+    }
+
+    if (scrollPos >= childProgressiveSizes.current[childProgressiveSizes.current.length - 1]) {
+      // Check end
+      return childProgressiveSizes.current.length - 1;
+    }
+
+    return findIndexRecursive(scrollPos, 0, childProgressiveSizes.current.length - 1);
+  };
+
+  const getIndexFromScrollPosition = (scrollPos: number) => {
+    if (!getItemSize) {
+      return Math.round(scrollPos / itemSize);
+    }
+
+    return getIndexFromSizeArray(scrollPos);
+  };
+
+  const calculateTotalSize = () => {
+    if (!getItemSize) {
+      return itemSize * childArray.length;
+    }
+
+    // Time for custom size calcs
+    return childProgressiveSizes.current[childArray.length - 1];
+  };
+
+  const calculateBefore = () => {
+    if (!getItemSize) {
+      // The missing items from before virtualization starts height
+      return virtualizerStartIndex * itemSize;
+    }
+
+    if (virtualizerStartIndex <= 0) {
+      return 0;
+    }
+
+    // Time for custom size calcs
+    return childProgressiveSizes.current[virtualizerStartIndex - 1];
+  };
+
+  const calculateAfter = () => {
+    if (childArray.length === 0) {
+      return 0;
+    }
+    const lastItemIndex = Math.min(virtualizerStartIndex + virtualizerLength, childArray.length - 1);
+    if (!getItemSize) {
+      // The missing items from after virtualization ends height
+      const remainingItems = childArray.length - lastItemIndex - 1;
+
+      return remainingItems * itemSize;
+    }
+
+    // Time for custom size calcs
+    return childProgressiveSizes.current[childArray.length - 1] - childProgressiveSizes.current[lastItemIndex];
+  };
+
+  const generateRows = (): ReactNode[] => {
+    if (childArray.length === 0) {
+      /* Nothing to virtualize */
+
+      return [];
+    }
+
+    const actualIndex = Math.max(virtualizerStartIndex, 0);
```

Comment:
This is future looking so don't block merging this PR on it, but I think we should look into using CSS containment as a performance optimization. Browser support is not 100% for our support matrix but since it's CSS it should degrade gracefully for browsers that don't support the required CSS while giving modern browsers better performance.

- [contain](https://developer.mozilla.org/en-US/docs/Web/CSS/contain)
- [content-visibility](https://developer.mozilla.org/en-US/docs/Web/CSS/content-visibility)
- [contain-intrinsic-size](https://developer.mozilla.org/en-US/docs/Web/CSS/contain-intrinsic-size#:~:text=The%20contain-intrinsic-size%20property%20allows%20authors%20to%20specify%20an,length%20when%20the%20element%20is%20skipping%20its%20contents.)

Reply:
100% agree, we were looking at optimizing the MADS dashboard with this since some apps have like 50 charts on them at times and reducing layout recalc could be big. Combine that with virtualization so things aren't rendered beyond the fold would be really nice.  I think the main thing to be careful with here is size. If host applications use this and expect child element sizing to impact other elements it could create unforeseen issues.